framework/src/systems/SystemBase.C

```diff
@@ -1352,6 +1357,26 @@ SystemBase::name() const
   return system().name();
 }
 
+void
+SystemBase::initSolutionStates()
+{
+  // Need the default (current) solution state
+  unsigned int state = 0;
+
+  // May need older states from the variables (requested in kernels, bcs, etc)
+  for (const auto & var : getVariables(/* tid = */ 0))
```

Comment:
I was thinking to let the executioner to drive what states we need. Kernels, etc. should error out if required state does not exist. In this way, we can get rid of integrity check currently in executioner. `FEProblemBase::transient` should be set properly based on executioner.

Reply:
I would be okay with this for the NL system, but not the aux system. There exist cases where you don't need older values in aux but you're running with a transient. In which case, there is no need to init the old aux vectors.

Reply:
This sounds to me a further optimization. Typically aux system does not cause any trouble because it is much smaller. I prefer a simple design here, i.e. if NL system has old, let aux system have it too.

Reply:
I'm inclined to agree with @YaqiWang here. Let the `Executioner` drive

Reply:
>This sounds to me a further optimization. Typically aux system does not cause any trouble because it is much smaller. I prefer a simple design here, i.e. if NL system has old, let aux system have it too.

I just want to mention in NDA, the aux is pretty big